### PR TITLE
Feature/readonlymode 427

### PR DESCRIPTION
**Pull request by hg-ms**
detail #427

### CODE REVIEW

File: persistence/persistence/src/main/java/one/microstream/persistence/exceptions/PersistenceExceptionStorerDeactivated.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * This exception indicates that the storage is currently in a read only mode
+ * that denies the usage of {@link one.microstream.persistence.types.PersistenceStorer} methods.
+ * 
+ */
+@SuppressWarnings("serial")
+public class PersistenceExceptionStorerDeactivated extends PersistenceException
+{
+	@Override
+	public String getMessage()
```
Comment (fh-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/exceptions/PersistenceExceptionStorerDeactivated.java:34]: Not sure if this produces the right output, getThrowingMethodName returns the first method name of the stack, meaning the method in the user code which has called a store method.

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerCreatorDeactivateAble.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import one.microstream.reference.ObjectSwizzling;
+import one.microstream.util.BufferSizeProviderIncremental;
+import one.microstream.util.logging.Logging;
+
+/**
+ * PeristenceStorer creator that creates {@link one.microstream.persistence.types.PersistenceStorerDeactivateAble#PersistenceStorerDeactivateAble PersistenceStorerDeactivateAble}
+ * instances.
+ * 
+ */
+public class PersistenceStorerCreatorDeactivateAble<D> implements PersistenceStorer.Creator<D>
```
Comment (fh-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerCreatorDeactivateAble.java]: We should go with Deactivatable as suffix here.

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+ */
+
+/**
+ * A {@link one.microstream.persistence.types.PersistenceStorer PersistenceStorer} implementation that allows
+ * switching between the supplied {@code PersistenceStorer} instance and a
+ * {@link one.microstream.persistence.types.PersistenceStorerDeactivated PersistenceStorerDeactivated}
+ * instance.
+ *
+ */
+public class PersistenceStorerDeactivateAble implements PersistenceStorer {
```
Comment (fh-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java]: We should go with Deactivatable as suffix here.

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	///////////////////////////////////////////////////////////////////////////
+	// methods //
+	////////////
+	
+	/**
+	 * Enable or disable writing support.
+	 * 
+	 * @param enabledWrites true to enable writing, false to disable.
+	 */
+	public void enableWrites(final boolean enabledWrites)
```
Comment (fh-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java]: We should go with getter/setter style naming, here setWriteEnabled(boolean), and also add a getter (isWriteEnabled)

**Review by fh-ms — CHANGES_REQUESTED** [review]
As we talked about, an idea would be to introduce a registry interface for the deactivatable persistence storers and hand this one to the creator instead of the controller.

File: persistence/persistence/src/main/java/one/microstream/persistence/exceptions/PersistenceExceptionStorerDeactivated.java (hunk 1 above)
Comment (hg-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/exceptions/PersistenceExceptionStorerDeactivated.java:34]: Is’s working as intended, the exeption is always thrown from a method in the ‘PersistenceStorerDeactivated‘ class.  The zeroth element of the stack trace is the method where throwable was created and thrown. Nevertheless I’ll fix as potential Index out of bounds error as the VM is allowed to return a zero sized stack trace if the VM has no stack strace.

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerCreatorDeactivateAble.java (hunk 2 above)
Comment (hg-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerCreatorDeactivateAble.java]: ok, done

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java (hunk 3 above)
Comment (hg-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java]: ok, done

File: persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java (hunk 4 above)
Comment (hg-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/types/PersistenceStorerDeactivateAble.java]: ok, done

File: persistence/persistence/src/main/java/one/microstream/persistence/types/StorerModeController.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+		this.registry = newRegistry;
+		logger.trace("Active storers after cleanup: {}", this.registry.size());
+	}
+	
+	/**
+	 * Switch all registered storers to the writing mode.
+	 * 
+	 * @return true if writing is enabled, otherwise false.
+	 */
+	public synchronized boolean enableWriting()
```
Comment (fh-ms) [inline persistence/persistence/src/main/java/one/microstream/persistence/types/StorerModeController.java]: To be consistent, getter/setter style methods would be best here as well.

Comment (hg-ms) [reply persistence/persistence/src/main/java/one/microstream/persistence/types/StorerModeController.java]: You're right. Unified API.

**Review by fh-ms — APPROVED** [review]

**Pull request merged**